Add SOCI index documentation (#877)
awslabs/soci-snapshotter: docs/index.md

> # SOCI Index

Reviewer: nit: Most of our docs contain a TOC at the top. Can you generate one using `doctoc .`?

Author: Added. PTAL.

## Index Conceptual Data Model

The high-level components of the SOCI index include:

> * [Index Manifest](#index-manifest) - an OCI image manifest.

Reviewer 2: Nit:
```suggestion
* [SOCI Index Manifest](#index-manifest) - an OCI image manifest describing the content of the SOCI Index.
```

Author: Applied.

* [zTOC](#ztoc) - a table of contents for compressed data.

### Index Manifest

The SOCI index manifest contains a subject reference to the image and a list of zTOCs.

* `Subject` _[descriptor](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc5/descriptor.md)_

    Reference to the image from which the index was generated.

> * `zTOCs` _array of [zTOC](#ztoc)_

Reviewer 2: Nit: this is technically descriptors of zTOCs.

![manifest-concept](images/soci-index-manifest-conceptual-data-model.drawio.svg)

### zTOC

A compression table of contents, or zTOC, is composed of 2 parts:

1. TOC - a table of contents containing file metadata and the file's offset
in the decompressed TAR.
1. zInfo - a collection of checkpoints of the state of the compression
engine at various points in the layer. Also referred to as compression info.

#### TOC

The table of contents contains the file metadata for each file in the TAR
needed to lazy load file content.

File metadata such as:

* Name
* Type

> * Offset

Reviewer 2: I think we might want to separate this from the rest of the data because this is the one piece that isn’t “normal” linux filesystem metadata.

It is probably also worth clarifying that this is the Offset of the file in the uncompressed TAR file of the layer.

* Size
* Permissions
* Timestamp

#### zInfo

Each checkpoint in the collection contains the state needed

> for the decoder to be able to decode a region of the layer TAR.

Reviewer 2: Nit:
```suggestion
to decompress a specific region of the layer TAR.
```

# SOCI Index

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Conceptual Data Model](#conceptual-data-model)
  - [SOCI Index Manifest](#soci-index-manifest)

>     - [zTOC](#ztoc)

Reviewer 2: nit: something got weird with this indentation and it looks like zTOC is a subsection of SOCI Index Manifest.

Author: Fixed in latest.

                "com.amazon.soci.image-layer-mediaType": "application/vnd.oci.image.layer.v1.tar+gzip"
            },
            "digest": "sha256:7ae9f4713a78a4895a7a522117c552a519d8289eecc446820b1b528cf9257b82",
            "mediaType": "application/octet-stream",
            "size": 841192
        },
        {
            "annotations": {
                "com.amazon.soci.image-layer-digest": "sha256:bb157f88c818b56b154696493271c0d7a2704ed253fc2111d0a96fa89bf44984",
                "com.amazon.soci.image-layer-mediaType": "application/vnd.oci.image.layer.v1.tar+gzip"
            },
            "digest": "sha256:715ee958c9cff9bca859e940cb05a3937bd2a6d2570fec910602df095339b067",
            "mediaType": "application/octet-stream",
            "size": 1086672
        }
    ],
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "schemaVersion": 2,
    "subject": {
        "digest": "sha256:ab5a62720aa71c964aae400825284ca3e8229934928394e68a06e395a225faae",
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 2752
    }
}
```

> ### Example Serialized zTOC

Reviewer: nit: I think we should have a note that mentions that this a _human readable_ format the resembles a `zTOC`, with some parts like the compression checkpoint information redacted. The actual `zTOC` is serialized as a binary file.

Author: Added in latest revision.
